M2/Macaulay2/packages/Varieties/SheafMaps.m2

```diff
@@ -0,0 +1,890 @@
+    and condition(G === sheaf(X, target matrix f),
+	"target of the sheaf map does not match the target of the underlying matrix")
+    and condition(F  == sheaf(X, source matrix f),
+	"source of the sheaf map does not match the source of the underlying matrix")
+    and condition(d >= min flatten degrees F, -- maybe not strictly necessary
+	"expected the degree of the sheaf map to be at least as high as the degrees of the source")
+    and condition(try ( isWellDefined map(module G, truncate(d, module F), matrix f) ) else false,
+	"expected the matrix to induce a map between a truncation of the underlying modules")
+    )
+
+-- basic methods
+source  SheafMap := CoherentSheaf => f -> f.source
+target  SheafMap := CoherentSheaf => f -> f.target
+variety SheafMap := Variety       => f -> f.source.variety
+ring    SheafMap := Ring          => f -> f.map.ring
+matrix  SheafMap := Matrix => opts -> f -> f.map
+degree  SheafMap := ZZ => f -> f.degree
+
+ker     SheafMap := CoherentSheaf => opts -> phi -> (sheaf ker matrix phi)
+image   SheafMap := CoherentSheaf => phi -> (sheaf image matrix phi)
+coimage SheafMap := CoherentSheaf => phi -> (sheaf coimage matrix phi)
+coker   SheafMap := CoherentSheaf => phi -> (sheaf coker matrix phi)
+
+-- TODO: is sheafMap sufficient here, or should we specify source/target/degree?
+cover   SheafMap := SheafMap => f -> sheafMap cover   matrix f
+ambient SheafMap := SheafMap => f -> sheafMap ambient matrix f
+super   SheafMap := SheafMap => f -> sheafMap super   matrix f
+
+-- TODO: add SheafMap == SheafMap
+SheafMap == ZZ := Boolean => (f, z) -> image f == z
+ZZ == SheafMap := Boolean => (z, f) -> image f == z
+
+isIsomorphism SheafMap := Boolean => f -> ker f == 0 and coker f == 0
+
+-- composition
+SheafMap * SheafMap := SheafMap => (phi, psi) -> (
+    d := degree phi; e := degree psi;
+    if d >= e
+    then map(target phi, source psi, matrix phi * truncate(d, matrix psi))
+    else map(target phi, source psi, truncate(d, matrix phi) * truncate(d, matrix psi), d)
+    )
+
+-- printing
+-- TODO: use abbreviations for source and target
+expression SheafMap := Expression => f -> (
+    if (s := f.map) == 0 then expression 0
+    else MapExpression { target f, source f, s })
+net              SheafMap :=      net @@ expression
+texMath          SheafMap :=  texMath @@ expression
+toString         SheafMap := toString @@ expression
+toExternalString SheafMap := toString @@ describe
+
+-----------------------------------------------------------------------------
+-- isLiftable, lift
+-----------------------------------------------------------------------------
+--general method: checks whether a map phi is in the image of
+--the map Hom(eta,target phi)
+isLiftable = method()
+isLiftable(Matrix, Matrix) := (phi, eta) -> (
+    newPhi := homomorphism' phi;
+    deta := Hom(eta, target phi);
+    (image newPhi / coker deta) == 0)
+
```

Comment:
The name ```deta``` seems strange.  What does it mean?

Reply:
@kellerlv do you remember?